### PR TITLE
ReactUtils – find by type – refactor using generic

### PR DESCRIPTION
**Pull request by marcalexiei**
## Description

In order to reduce the `any` types in the code I change the implementation of `findAllByType` and `findChildByType` to use generic to better infer component props in the return value:

Screenshot 2023-01-06 at 13 55 24

To achieve this now these two functions only accepts `React.ComponentType` parameter instead of `string` in order to retrieve the `Props`. 
All invocation of the function have been changed. Most of them was using `Component.displayName` so since `getDisplayName` accounts for both `string` and `React.ComponentType` this should't be a problem.

---

While doing this I encountered an error in `appendOffsetOfLegend` function which was comparing `verticalAlign` `Legend` props with `"center"`.  It looks like it isn't a valid value so I changed it to `"middle"` since `VerticalAlignmentType ` values are 
```ts
'top' | 'bottom' | 'middle'
```

[Commit reference](https://github.com/recharts/recharts/commit/5a4a9874f2b6f20877cc73b4338dbfd06c2745fe)

---

Let me know if you prefer to have two separate PR for the issue and the refactor.

## Related Issue
N/A

## Motivation and Context
Improve code navigation and reference

## How Has This Been Tested?

`npm run build`
`npm run test`

## Screenshots (if appropriate):
N/A

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

File: src/util/ReactUtils.ts
Comment (nikolasrieble) [inline src/util/ReactUtils.ts:101]: Nit: Lets not use abbreviations. Instead, please write out the fields.
```suggestion
  Component extends React.ComponentType,
```

File: src/util/ReactUtils.ts
Comment (nikolasrieble) [inline src/util/ReactUtils.ts:102]:
```suggestion
  ReturnComponent = React.DetailedReactHTMLElement<React.ComponentProps<Component>, HTMLElement>,
```

File: .vscode/settings.json
Comment (nikolasrieble) [inline .vscode/settings.json:1]: I do not think we need to add vscode settings to recharts.

**Review by nikolasrieble — CHANGES_REQUESTED** [review]
Please remove the vscode settings. Other than that, the change looks good! Thank you for the refactoring.

File: .vscode/settings.json
Comment (ckifer) [reply .vscode/settings.json:1]: +1, haven't looked at the rest of the PR yet but editor settings should be in our gitignore. Each devs editor settings are their own and we'll let eslint/prettier take care of formatting

File: src/util/ChartUtils.ts
Comment (nikolasrieble) [inline src/util/ChartUtils.ts:389]: As this is a clear bug fix, I think we are good to keep this change. Center was never an option :)

**Comment by marcalexiei** [comment]
I have renamed generic to avoid abbreviation usage and added `.vscode` folder to .gitignore

**Comment by marcalexiei** [comment]
I would consider to merge and not squash this PR to have the fix clearly visible in the commit history. :-)

File: src/util/ChartUtils.ts
Comment (ckifer) [reply src/util/ChartUtils.ts:389]: +1, nice find

File: src/util/ReactUtils.ts
Comment (ckifer) [inline src/util/ReactUtils.ts:98]: Can we update this comment? It looks like now `type`  can only be `ComponentType | ComponentType[]` which does not include string

Also does this work with a Function component?

File: src/util/ReactUtils.ts
Comment (ckifer) [inline src/util/ReactUtils.ts:125]: Same here, comment needs updated. String no longer accepted and curious if function components get desired results.

File: src/util/ReactUtils.ts
Comment (ckifer) [inline src/util/ReactUtils.ts:133]: I would rather avoid the cast to `never` here if we could

Instead of creating `DetailedElement` can we just make the return type
`React.DetailedReactHTMLElement<React.ComponentProps<ComponentType>, HTMLElement>`

so the the cast to never can be removed?

Or if we don't want the return type to be so long to do something like
```ts
type DetailedElement<ComponentType extends React.ComponentType> = React.DetailedReactHTMLElement<
  React.ComponentProps<ComponentType>,
  HTMLElement
>;
...
export function findChildByType<ComponentType extends React.ComponentType>(
  children: ReactNode[],
  type: ComponentType | ComponentType[],
): DetailedElement<ComponentType> {
```

**Review by ckifer — CHANGES_REQUESTED** [review]

File: src/util/ReactUtils.ts
Comment (marcalexiei) [reply src/util/ReactUtils.ts:98]: > Can we update this comment?

Sure, I haven't noticed that earlier. 😅

> Also does this work with a Function component?

`ComponentType` includes both component types
```ts
type ComponentType<P = {}> = ComponentClass<P> | FunctionComponent<P>;
```

File: src/util/ReactUtils.ts
Comment (marcalexiei) [reply src/util/ReactUtils.ts]: I removed the return type alongside the `never` casting.
TS infers all correct types from `findAllByType`.

**Review by ckifer — APPROVED** [review]
LGTM, thanks for doing this!

Will merge this with a merge commit. 

I'm going to have to mull on if to merge this yet or wait until after I release 2.3.0 - we're going to need to adopt a different branching strategy at some point I think (@nikolasrieble)

**Comment by marcalexiei** [comment]
> I'm going to have to mull on if to merge this yet or wait until after I release 2.3.0

In the end this PR contains a bugfix and mostly refactor, so I think is ok to merge since 2.3.0 will contain many bugfixes.
The only notable change is on [`src/util/DetectReferenceElementsDomain.ts`](https://github.com/recharts/recharts/pull/3152/files#diff-55d36b01ae78f58b3f857d1cd10b1f20a33001bd96f5de913f5672cc80ba71b2R18) where I changed:
```ts
const elements = lines.concat(dots);
```
to
```ts
const elements = [...lines, ...dots];
```
That shouldn't cause any issue.

> we're going to need to adopt a different branching strategy at some point I think

When the work on the 3.0 starts you can create a 2.x branch where the fix can be backported from master. 
Meanwhile master will become the house of 3.x version code. This is the simplest approach I can think of.

**Pull request merged**
